lib/paseto/public.rb

```diff
@@ -1,54 +1,84 @@
-      new(Paseto.decode64(encoded_key), footer)
-    end
+        def self.decode64(encoded_key)
+          new(Paseto.decode64(encoded_key))
+        end
 
-    def initialize(private_key, footer = nil)
-      @signing_key = RbNaCl::SigningKey.new(private_key)
-      @verify_key = @signing_key.verify_key
-      @footer = footer
-    end
+        def initialize(key)
+          @key = key
+          @nacl = RbNaCl::SigningKey.new(key)
+        end
 
-    def sign(payload)
-      data = encode_message(payload)
-      # Sign a message with the signing key
-      signature = @signing_key.sign(data)
+        def public_key
+          PublicKey.new(@nacl.verify_key.to_bytes)
+        end
 
-      Paseto::Message.new(HEADER, data + signature, @footer).to_message
-    end
+        def encode64
+          Paseto.encode64(@key)
+        end
 
-    def verify(message)
-      raise Paseto::BadHeaderError.new('Invalid message header.') unless message.start_with?(HEADER)
+        attr_reader :nacl
+      end
 
-      computed_msg = Paseto.validate_and_remove_footer(message, @footer)
-      decoded_payload = Paseto.decode64(computed_msg[10..-1]);
+      class PublicKey
+        def self.decode64(encoded_key)
+          new(Paseto.decode64(encoded_key))
+        end
 
-      decoded_message = decoded_payload[0..-(SIGNATURE_BYTES + 1)]
-      signature = decoded_payload[-SIGNATURE_BYTES..-1]
+        def initialize(key)
+          @key = key
+          @nacl = RbNaCl::VerifyKey.new(key)
+        end
 
-      raise BadMessageError.new('Unable to process message') if decoded_message.nil? || signature.nil?
+        def encode64
+          Paseto.encode64(@key)
+        end
 
-      begin
-        @verify_key.verify(signature, decoded_message)
-      rescue RbNaCl::BadSignatureError
-        false
+        attr_reader :nacl
       end
-    end
 
-    private
+      def self.sign(message, key, footer = nil)
+        data = encode_message(message, footer)
+        # Sign a message with the signing key
+        signature = key.nacl.sign(data)
+
+        Paseto::Token.new(HEADER, message + signature, footer).to_message
+      end
 
-    def encode_message(message)
-      Paseto.pre_auth_encode(HEADER, message, @footer)
+      def self.verify(token, key, footer = nil)
+        parsed = Paseto.parse_raw_token(token, HEADER, footer)
+
+        decoded_message = parsed.payload[0..-(SIGNATURE_BYTES + 1)]
+        signature = parsed.payload[-SIGNATURE_BYTES..-1]
+
+        raise BadMessageError.new('Unable to process message') if decoded_message.nil? || signature.nil?
+
+        begin
+          key = key.public_key if key.is_a? SecretKey
+          data = encode_message(decoded_message, footer)
+          key.nacl.verify(signature, data)
+          decoded_message
+        rescue RbNaCl::BadSignatureError
+          raise AuthenticationError.new('Token signature invalid')
+        end
+      end
+
+      private
+
+      def self.encode_message(message, footer)
```

Comment:
If this needs to be private, it has to be defined in 

    class << self
    end

but this is also a code smell resulting from relying on class methods to much.